Kernel/WorkQueue.h

```diff
@@ -42,52 +42,33 @@ class WorkQueue {
 
     WorkQueue(const char*);
 
-    void queue(void (*function)(void*), void* data = nullptr, void (*free_data)(void*) = nullptr)
+    void queue(Function<void(void*)> callback)
     {
-        auto* item = new WorkItem; // TODO: use a pool
-        item->function = function;
-        item->data = data;
-        item->free_data = free_data;
-        do_queue(item);
-    }
-
-    template<typename Function>
-    void queue(Function function)
-    {
-        auto* item = new WorkItem; // TODO: use a pool
-        item->function = [](void* f) {
-            (*reinterpret_cast<Function*>(f))();
-        };
-        if constexpr (sizeof(Function) <= sizeof(item->inline_data)) {
-            item->data = new (item->inline_data) Function(move(function));
-            item->free_data = [](void* f) {
-                reinterpret_cast<Function*>(f)->~Function();
-            };
-
-        } else {
-            item->data = new Function(move(function));
-            item->free_data = [](void* f) {
-                delete reinterpret_cast<Function*>(f);
-            };
-        }
-        do_queue(item);
+        do_queue(WorkItem::create(move(callback), nullptr, nullptr));
     }
 
 private:
-    struct WorkItem {
-        IntrusiveListNode m_node;
-        void (*function)(void*);
-        void* data;
-        void (*free_data)(void*);
+    class WorkItem {
+    public:
+        static NonnullOwnPtr<WorkItem> create(Function<void(void*)>, Function<void(void*)>, void*);
+        void callback() const;
+        void free_data_callback() const;
+
+    private:
+        WorkItem(Function<void(void*)>, Function<void(void*)>, void*);
+        Function<void(void*)> m_callback;
+        Function<void(void*)> m_free_data_callback;
+        void* m_data { nullptr };
```

Comment:
I'm not sure what the point of `m_free_data_callback` and `m_data` is if you use `Function<>`. The reason I used these with the c function pointers is simply to support lambdas, and to support them efficiently. And it gives you the flexibility to use lambdas or static function handlers without any allocation overhead if you don't have a need for it. To acoomodate lambdas the data pointer is set to the lambda instance and the free callback is set to the deleter of that object. It also allowed for in-place lambdas to save on unnecessary memory allocations, so if you only capture one pointer such as `this` it could be done with one allocation less.

To be honest I don't really see any advantage here, it's more overhead and not as nice to use. I modeled this after the deferred functions in `Processor`, it's lightweight and supports lambdas just fine. Note that `Function` currently even uses a `RefPtr` internally which makes it even more heavy than it should be.

Reply:
Also, I had plans to use a `WorkItem` pool so that we don't it `kmalloc` every time we want to queue something.

Reply:
I don't see where `Function` uses a `RefPtr`, but I do see your point, the code should be lightweight if we happen to do lots of IO operations with it. Still, all the `reinterpret_cast`s look suspicious, but maybe we should accommodate them.

Reply:
Sorry, my bad, a `OwnPtr`:
https://github.com/SerenityOS/serenity/blob/0d8c9024eec12d57270cca1cf6bfc7d03c0ca229/AK/Function.h#L117

Either way, that shouldn't really matter for the scope of this PR, it was just one thing I don't like about `Function` and have been meaning to fix it.